src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,669 @@
+        }
+
+        synchronized (lock) {
+            Exception lastException = null;
+            if (!checkSpiNonNull(candidate)) {
+                throw new RuntimeException("Unexpected Error: candidate is null!");
+            }
+            Delegate currOne = candidate;
+            try {
+                while (true) {
+                    try {
+                        Object result = (isDeriveData) ?
+                                currOne.spi().engineDeriveData(derivationSpec) :
+                                currOne.spi().engineDeriveKey(algorithm,
+                                        derivationSpec);
+                        // found a working KDFSpi
+                        this.theOne = currOne;
+                        return result;
+                    } catch (Exception e) {
+                        if (lastException == null) {
+                            lastException = e;
+                        }
+                        // try next one if available
+                        assert serviceIterator != null : "serviceIterator was null";
+                        currOne = getNext(serviceIterator, kdfParameters);
+                    }
+                }
+            } catch (InvalidAlgorithmParameterException e) {
+                throw e; // getNext reached end and have seen IAPE
+            } catch (NoSuchAlgorithmException e) {
+                // getNext reached end without finding an implementation
+                throw new InvalidAlgorithmParameterException(lastException);
+            }
+        }
+    }
+
+    private static Delegate getNext(Iterator<Service> serviceIter,
+                                    KDFParameters kdfParameters)
+            throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        // fetch next one if available
+        boolean hasOne = false;
+        while (serviceIter.hasNext()) {
+            Service s = serviceIter.next();
+            Provider prov = s.getProvider();
+            if (!JceSecurity.canUseProvider(prov)) {
+                // continue to next iteration
+                continue;
+            }
+            hasOne = true;
+            try {
+                Object obj = s.newInstance(kdfParameters);
+                if (!(obj instanceof KDFSpi)) {
+                    if (!skipDebug && pdebug != null) {
+                        pdebug.println(
+                                "obj was not an instance of KDFSpi (should not "
+                                + "happen)");
+                    }
+                    continue;
+                }
+                return new Delegate((KDFSpi) obj, prov);
+            } catch (NoSuchAlgorithmException nsae) {
+                // continue to the next provider
+                if (!skipDebug && pdebug != null) {
+                    pdebug.println(
+                            "The evaluated provider does not support the "
+                            + "specified algorithm and parameters");
```

Comment:
`nsae.printStackTrace(pdebug.getPrintStream());`

Reply:
You can keep the existing `pdebug.println` line. Maybe put a line before line 622 as well. When a program terminated with an exception, it's usually OK to just dump the exception there without a comment since it's not handled. In this case, we actually handle the exception, and it's better to add a comment on how we handled it.